src/lib/mina_base/zkapp_command.ml

```diff
@@ -1524,15 +1532,48 @@ struct
       zkapp.verification_key
     in
     let tbl = Account_id.Table.create () in
-    let%map () =
-      Call_forest.fold t.account_updates ~init:(Some ()) ~f:(fun acc p ->
-          let%bind _ok = acc in
+    let%map (_all_succeeded : _ Account_id.Map.t) =
+      Call_forest.fold t.account_updates
+        ~init:
+          ((* keep track of the verification keys that have been set so far
+            * during this transaction -- the option tracks if we should stop
+            * early due to an error *)
+             Some
+             Account_id.Map.empty ) ~f:(fun acc p ->
+          let%bind vks_overrided = acc in
           let account_id = Account_update.account_id p in
+          let vks_overrided' =
```

Comment:
I believe it's 100% safe to always record the newly set verification key unconditionally (regardless of authorization), but would love someone to double check if I'm thinking through this properly.

Reply:
If it isn't safe, then the transaction will get rejected at that point, and we'll never end up using that vk :+1: